### PR TITLE
Implementing TKG AWS workload cluster Implementation with Acceptance Testing

### LINKED ISSUE #37: Ability to Provision TKGm Clusters on AWS using Terraform
### Is your feature request related to a problem? Please describe.

No

### Describe the solution you'd like

Currently it seems this plugin can only provision TKGm clusters via TMC using vSphere,  I would like to be able to provision clusters on other cloud providers,  most importantly AWS,  but I could see the same for other clouds.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Vasundharashukla**
1. **What this PR does / why we need it**: To support the provisioning of TKG AWS workload cluster.

2. **Which issue(s) this PR fixes**

        (optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged):

        Fixes #37 

3. **Additional information**
Adding Unit Tests and Acceptance Test for the same

4. **Special notes for your reviewer**

### CODE REVIEW

File: examples/usecases/usecase.tf [hunk 1]
```diff
@@ -6,12 +6,12 @@
...
     tanzu-mission-control = {
-      source = "vmware/tanzu-mission-control"
+      source = "vmware/dev/tanzu-mission-control"
     }
   }
 }
 
-// Create workspace
+# Create workspace
 resource "tanzu-mission-control_workspace" "create_workspace" {
```
Comment (gsadhani) [inline examples/usecases/usecase.tf:15]: why is there a mixed style (snake and with `-` separator) naming for many resources?

File: examples/usecases/usecase.tf [hunk 2]
```diff
@@ -23,13 +23,13 @@ resource "tanzu-mission-control_workspace" "create_workspace" {
...
 
-// Create cluster group
+# Create cluster group
 resource "tanzu-mission-control_cluster_group" "create_cluster_group" {
   name = "demo-cluster-group"
 }
 
-// Create Tanzu Mission Control attach cluster with k8s cluster kubeconfig provided
-// The provider would create the cluster entry and apply the deployment link manifests on to the k8s kubeconfig provided.
+# Create Tanzu Mission Control attach cluster with k8s cluster kubeconfig provided
```
Comment (gsadhani) [inline examples/usecases/usecase.tf:31]: nit: Attach a Tanzu Mission Control cluster with k8s cluster kubeconfig provided

File: examples/usecases/usecase.tf [hunk 3]
```diff
@@ -224,4 +224,87 @@ resource "tanzu-mission-control_cluster_node_pool" "create_node_pool" {
       storage_class = "gc-storage-profile"
     }
   }
+}
+
+# Create Tanzu Mission Control Tanzu Kubernetes Grid AWS workload cluster entry
+resource "tanzu-mission-control_cluster" "create_tkg_aws_cluster" {
```
Comment (gsadhani) [inline examples/usecases/usecase.tf:230]: would the resource name like `tkg_aws_cluster` be more appropriate?

File: internal/models/cluster/nodepool/nodepools_spec_tkg_aws.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	// The potential values could be found using cluster:options api.
+	InstanceType string `json:"instanceType,omitempty"`
+
+	// List of AZs to place the AWS nodes on. Please use this field to provision a nodepool for workload cluster on an attached TKG AWS management cluster.
+	// Please specify 1 AZ for a dev cluster and up to 3 AZs for production cluster.
+	NodePlacement []*VmwareTanzuManageV1alpha1ClusterNodepoolTKGAWSNodePlacement `json:"nodePlacement"`
+
+	// The nodes private CIDR block if we are creating a new VPC.
+	// This is no longer used - Please dont use this deprecated field.
+	PrivateCidr string `json:"privateCidr,omitempty"`
```
Comment (gsadhani) [inline internal/models/cluster/nodepool/nodepools_spec_tkg_aws.go:32]: nit: `PrivateCIDR`

Comment (gsadhani) [reply internal/models/cluster/nodepool/nodepools_spec_tkg_aws.go:32]: Also, since it is deprecated should we remove it?

File: internal/models/cluster/tkgaws/tkgaws_spec.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+package tkgawsmodel
+
+import (
+	"github.com/go-openapi/swag"
+)
+
+// VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsSpec TKG AWS cluster spec.
+//
+// swagger:model vmware.tanzu.manage.v1alpha1.cluster.infrastructure.tkgaws.Spec
+type VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsSpec struct {
```
Comment (gsadhani) [inline internal/models/cluster/tkgaws/tkgaws_spec.go:15]: `VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsSpec ` -> `VmwareTanzuManageV1alpha1ClusterInfrastructureTKGAWSSpec`

File: internal/models/cluster/nodepool/nodepools_spec_tkg_aws.go (hunk 4 above)
Comment (gsadhani) [inline internal/models/cluster/nodepool/nodepools_spec_tkg_aws.go:32]: `PrivateCIDR`.

Also should we remove this as it is deprecated.

File: internal/models/cluster/tkgaws/tkgaws_cluster_network.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+package tkgawsmodel
+
+import (
+	"github.com/go-openapi/swag"
+)
+
+// VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsClusterNetwork Network information for the cluster.
+//
+// swagger:model vmware.tanzu.manage.v1alpha1.cluster.infrastructure.tkgaws.ClusterNetwork
+type VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsClusterNetwork struct {
```
Comment (gsadhani) [inline internal/models/cluster/tkgaws/tkgaws_cluster_network.go:15]: `VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsClusterNetwork` -> `VmwareTanzuManageV1alpha1ClusterInfrastructureTKGAWSClusterNetwork`

File: internal/models/cluster/tkgaws/tkgaws_subnet.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+//
+// swagger:model vmware.tanzu.manage.v1alpha1.cluster.infrastructure.tkgaws.Subnet
+type VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsSubnet struct {
+
+	// AWS availability zone e.g. us-west-2a
+	AvailabilityZone string `json:"availabilityZone,omitempty"`
+
+	// CIDR for AWS subnet.
+	// This CIDR block must be in the range of AWS VPC CIDR block.
+	CidrBlock string `json:"cidrBlock,omitempty"`
```
Comment (gsadhani) [inline internal/models/cluster/tkgaws/tkgaws_subnet.go:22]: `CIDRBlock`

File: internal/resources/cluster/config_test.go [hunk 8]
```diff
@@ -63,6 +63,80 @@ const testDataSourceAttachClusterScript = `
...
+const testTKGmAWSClusterScript = `
+ 	resource {{.ResourceName}} {{.ResourceNameVar}} {
+ 	  management_cluster_name = "{{.ManagementClusterName}}"
+ 	  provisioner_name        = "{{.ProvisionerName}}"
+ 	  name                    = "{{.Name}}"
+
+      spec {
+		cluster_group = "default"
+		tkg_aws {
+		settings {
```
Comment (gsadhani) [inline internal/resources/cluster/config_test.go:75]: nit: indentation

File: internal/resources/cluster/resource_cluster.go [hunk 9]
```diff
@@ -161,6 +163,12 @@ func constructSpec(d *schema.ResourceData) (spec *clustermodel.VmwareTanzuManage
 		spec.ProxyName = v.(string)
 	}
```
Comment (gsadhani) [inline internal/resources/cluster/resource_cluster.go:165]: Should we throw an error if none of the TKG-AWS, TKGS-vSphere or TKG-vSphere specs is configured?

File: internal/resources/cluster/tkgaws/resource_tkg_aws.go [hunk 10]
```diff
@@ -0,0 +1,967 @@
...
+
+	return []interface{}{flattenProviderNetwork}
+}
+
+var tkgAWSSubnets = &schema.Schema{
+	Type:        schema.TypeList,
+	Description: "Optional list of subnets used to place the nodes in the cluster",
+	Optional:    true,
+	Elem: &schema.Resource{
+		Schema: map[string]*schema.Schema{
```
Comment (shobha2626) [inline internal/resources/cluster/tkgaws/resource_tkg_aws.go:404]: Please add computed to the fields which gets autofilled. We also have certain fields like vpc_id and subnet id right, please handle those fields as well

File: internal/resources/cluster/tkgaws/resource_tkg_aws.go [hunk 11]
```diff
@@ -0,0 +1,967 @@
...
+	flattenProviderNetwork[subnetsKey] = sns
+
+	if provider.Vpc != nil {
+		flattenProviderNetwork[vpcKey] = flattenTKGAWSVPC(provider.Vpc)
+	}
+
+	return []interface{}{flattenProviderNetwork}
+}
+
+var tkgAWSSubnets = &schema.Schema{
```
Comment (shobha2626) [inline internal/resources/cluster/tkgaws/resource_tkg_aws.go:399]: Let's also mention the order to be followed for the subnet blocks in the terraform script

**Review by shobha2626 — CHANGES_REQUESTED** [review]
Please rebase the PR

File: internal/models/cluster/nodepool/nodepools_spec_tkg_aws.go (hunk 4 above)
Comment (Vasundharashukla) [reply internal/models/cluster/nodepool/nodepools_spec_tkg_aws.go:32]: @gsadhani the spec attributes int he model files are autogenerated using the go swagger tool, these variables are not manually named here. Not used these specs in our resources and examples, but can be removed from here too.

File: internal/models/cluster/tkgaws/tkgaws_spec.go (hunk 5 above)
Comment (Vasundharashukla) [reply internal/models/cluster/tkgaws/tkgaws_spec.go:15]: The spec is auto-generated using the go swagger tool, these variables are not manually named here.

File: internal/models/cluster/tkgaws/tkgaws_cluster_network.go (hunk 6 above)
Comment (Vasundharashukla) [reply internal/models/cluster/tkgaws/tkgaws_cluster_network.go:15]: Same as above

File: Makefile [hunk 12]
```diff
@@ -4,9 +4,9 @@
 default: build
 
 build:
-	go build -o bin/terraform-provider-tanzu-mission-control_v1.0.0
-	mkdir -p ~/.terraform.d/plugins/vmware/dev/tanzu-mission-control/1.0.0/darwin_amd64/
-	cp bin/terraform-provider-tanzu-mission-control_v1.0.0 ~/.terraform.d/plugins/vmware/dev/tanzu-mission-control/1.0.0/darwin_amd64/
+	go build -o bin/terraform-provider-tanzu-mission-control_v1.0.1
```
[the remark this answers is not in the document]
Comment (Vasundharashukla) [reply Makefile:7]: Yes, that was a miss. We change it to the original version.

File: internal/resources/cluster/tkgaws/resource_tkg_aws.go (hunk 10 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/tkgaws/resource_tkg_aws.go:404]: They have already been taken care.

File: internal/resources/cluster/tkgaws/resource_tkg_aws.go (hunk 11 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/tkgaws/resource_tkg_aws.go:399]: Where do you exactly want this to be mentioned? README?

Comment (Vasundharashukla) [reply internal/resources/cluster/tkgaws/resource_tkg_aws.go:399]: We can add in the guides as a **NOTE**

File: internal/models/cluster/tkgaws/tkgaws_subnet.go (hunk 7 above)
Comment (Vasundharashukla) [reply internal/models/cluster/tkgaws/tkgaws_subnet.go:22]: Same as above

File: internal/resources/cluster/tkgaws/resource_tkg_aws.go [hunk 13]
```diff
@@ -0,0 +1,967 @@
...
+				Optional:    true,
+			},
+			tkgAWSKey: {
+				Type:        schema.TypeList,
+				Description: "Nodepool config for tkg aws",
+				Optional:    true,
+				MaxItems:    1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						nodepoolAvailabilityZoneKey: {
```
Comment (shobha2626) [inline internal/resources/cluster/tkgaws/resource_tkg_aws.go:767]: Nodepool schema is again specific to the version of the AWS workload cluster.
image


Since we have `tkgAWSKey` as optional it should be fine

File: internal/resources/cluster/tkgaws/resource_tkg_aws.go [hunk 14]
```diff
@@ -0,0 +1,967 @@
...
+	}
+
+	flattenControlPlane[availabilityZonesKey] = controlPlane.AvailabilityZones
+	flattenControlPlane[instanceTypeKey] = controlPlane.InstanceType
+	flattenControlPlane[highAvailabilityKey] = controlPlane.HighAvailability
+
+	return []interface{}{flattenControlPlane}
+}
+
+var tkgAWSNodePoolInfo = &schema.Schema{
```
Comment (shobha2626) [inline internal/resources/cluster/tkgaws/resource_tkg_aws.go:730]: Add description

**Review by shobha2626 — CHANGES_REQUESTED** [review]
Please add the optional parameters for AWS workload cluster support

File: internal/resources/cluster/tkgaws/resource_tkg_aws.go (hunk 11 above)
Comment (shobha2626) [reply internal/resources/cluster/tkgaws/resource_tkg_aws.go:399]: README seems appropriate

File: internal/resources/cluster/resource_cluster.go (hunk 9 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:168]: Terraform takes care of the error handling when we do not populate any of the specs as those are required fields. The following error is seen: `POST request failed with status : 400 Bad Request, response: {"error":"Invalid value for management cluster name and/or provisioner name in request","code":3`

File: internal/resources/cluster/resource_cluster_test.go [hunk 15]
```diff
@@ -88,9 +95,9 @@ func testGetResourceClusterDefinition(t *testing.T, opts ...testAcceptanceOption
 			t.Skipf("KUBECONFIG env var is not set: %s", templateConfig.KubeConfigPath)
 		}
 
-	case tkgsCluster:
```
Comment (shobha2626) [inline internal/resources/cluster/resource_cluster_test.go:91]: We need all the three cases right?

File: internal/resources/cluster/tkgaws/resource_tkg_aws.go (hunk 11 above)
Comment (shobha2626) [reply internal/resources/cluster/tkgaws/resource_tkg_aws.go:399]: guides as a NOTE is also fine, lets capture it

File: examples/resources/cluster/resource_cluster_tkg_aws.tf [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+            pods {
+              cidr_blocks = "100.96.0.0/11" // Required
+            }
+
+            services {
+              cidr_blocks = "100.64.0.0/13" // Required
+            }
+          }
+          provider {
+            subnets {
```
Comment (shobha2626) [inline examples/resources/cluster/resource_cluster_tkg_aws.tf:27]: Capture the ordering of subnets in AWS (in doc section)

File: internal/resources/cluster/tkgaws/resource_tkg_aws.go [hunk 17]
```diff
@@ -0,0 +1,970 @@
...
+				Type:        schema.TypeString,
+				Description: "Specifies region of the cluster",
+				Required:    true,
+			},
+			versionKey: {
+				Type:        schema.TypeString,
+				Description: "Specifies version of the cluster",
+				Required:    true,
+			},
+			provisionerCredentialKey: {
```
[the remark this answers is not in the document]
Comment (shobha2626) [reply internal/resources/cluster/tkgaws/resource_tkg_aws.go:544]: Handled

File: internal/resources/cluster/resource_cluster_test.go (hunk 15 above)
Comment (shobha2626) [reply internal/resources/cluster/resource_cluster_test.go:91]: placed below

File: internal/resources/cluster/nodepools/resource_node_pool.go [hunk 18]
```diff
@@ -239,25 +463,6 @@ func flattenTkgServiceVsphere(tkgServiceVsphere *nodepoolmodel.VmwareTanzuManage
 	return []interface{}{flattenTkgServiceVsphereData}
 }
 
-func constructTkgServiceVsphere(data []interface{}) (tkgServiceVsphere *nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGServiceVsphereNodepool) {
```
Comment (shobha2626) [inline internal/resources/cluster/nodepools/resource_node_pool.go:242]: Why have we removed this construct function?

**Comment by shobha2626** [comment]
Rebase with main and squash the commits

File: internal/resources/cluster/nodepools/resource_node_pool.go (hunk 18 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/nodepools/resource_node_pool.go:242]: It is present [here.](https://github.com/vmware/terraform-provider-tanzu-mission-control/blob/acb97de32f9c5bcbed468231c3458ac6f7509eed/internal/resources/cluster/nodepools/resource_node_pool.go#L352) or else our tests would have been failing 😅

File: internal/models/cluster/nodepool/nodepools_spec.go [hunk 19]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright © 2021 VMware, Inc. All Rights Reserved.
+Copyright © 2022 VMware, Inc. All Rights Reserved.
```
Comment (shobha2626) [inline internal/models/cluster/nodepool/nodepools_spec.go:2]: Please revert this change

File: internal/resources/cluster/resource_cluster.go [hunk 20]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright © 2021 VMware, Inc. All Rights Reserved.
+Copyright © 2022 VMware, Inc. All Rights Reserved.
```
Comment (shobha2626) [inline internal/resources/cluster/resource_cluster.go:2]: Revert this change, 2022 is added to new files added

**Review by shobha2626 — CHANGES_REQUESTED** [review]

**Review by gsadhani — DISMISSED** [review]

**Review by shobha2626 — CHANGES_REQUESTED** [review]
LGTM

**Review by gsadhani — APPROVED** [review]
